protocols/request-response/src/lib.rs

```diff
@@ -677,7 +677,34 @@ where
         }
     }
 
-    fn on_dial_failure(&mut self, DialFailure { peer_id, .. }: DialFailure) {
+    // Removes the connection if it exists.
+    fn remove_connection(&mut self, connection_id: ConnectionId, peer: Option<PeerId>) {
+        if let Some(peer) = peer {
```

Comment:
I don't understand the benefit of doing this. I'd much rather always remove by `ConnectionId` and simply ignore the `PeerId`.

Perhaps we should change our internal data structure to index by `ConnectionId` always?

Reply:
> Perhaps we should change our internal data structure to index by ConnectionId always?

Generally yes that makes sense. I took a look at the API the request response exposes and there are a few methods that want to look connections by peer id and do not have access to the connection id. They are 

* is_connected
* is_pending_outbound
* is_pending_inbound
* try_send_request

I am not too familiar with the access patterns of the behavior and not sure if the trade offs make sense. If we change the internal structure to be indexed by connection id these methods above would be O(n) instead of O(1) and removing a connection state for a failed connection would be O(1) vs O(n).

My hunch is that the above methods are a more common and more critical code path than removing connections for failed dials/listens, so therefore we should leave the internal index by PeerId.

Or I could change it to keep two indexes but that seems more complicated that we might want. 

Thoughts? Happy to go in either direction.

Reply:
Lets go with O(n) for removing connections. I think there is an opportunity to build a data structure which can be used across all protocols for indexing some state `T` by both IDs but we don't have to build that in this PR.

Let me know if you are interested though :)

Reply:
For this PR I agree lets keep it simple. However a shared data structure across proctocols sounds very useful

Reply:
Do you mind opening an issue so we can track this?